### ISSUE TITLE
RSPEC-2386 was recognized as RSPEC-2223

### ISSUE THREAD

**Issue by chucklu** [SonarSource/sonar-dotnet#3636]
### Description
I have the following code, and it was recognized as RSPEC-2223 "Non-constant static fields should not be visible
"
```C#
 public class CustomHelper
    {
        public static List<string> MandatoryFields = new List<string>
        {
            "First name",
            "Last name"
        };
    }
```

I tried to fix as mentioned https://jira.sonarsource.com/browse/RSPEC-2223, just add readonly keyword
```C#
public class CustomHelper
    {
        public static readonly List<string> MandatoryFields = new List<string>
        {
            "First name",
            "Last name"
        };
    }
```
And you give me another issue, [Mutable, non-private fields should not be "readonly"](https://jira.sonarsource.com/browse/RSPEC-3887),  and this issue is not listed on official documents.
https://rules.sonarsource.com/csharp?search=mutable
It makes me confused, because the solution in RSPEC-3887 ask me to remove the keyword readonly.

And when I search mutable on official document, and I find [RSPEC-2386](https://jira.sonarsource.com/browse/RSPEC-2386). I  realized that you provide the wrong report for the above code. The correct report should be RSPEC-2386 instead of RSPEC-2223.

### Related information
 The sonarqube version we used is Community Edition, Version 8.4.2 (build 36762)

**Comment by pavel-mikula-sonarsource** [comment]
Hi @chucklu,

The S3887 doesn't ask you to remove the `readonly` keyword. It asks you to `Use an immutable collection or reduce the accessibility of the field(s).`. You should replace your `List<string>` with `string[]` or something like `ImmutableList<string>`. 

The whole situation is that
S2223 doesn't like that anybody can replace your list instance
S3887 doesn't like public+readonly on mutable collection
S2386 doesn't like public+static on mutable collection
So you'll actually get S3887 and S2386 if you do `public static readonly`

If you really need a mutable collection, than public static field is not what you should use for a clean code. That's what the rules are trying to prevent you from using.

Regarding rules website, there's probably just some indexing issue. When I search like this, I can find the rule S3887: 
https://rules.sonarsource.com/csharp?search=readonly

**Comment by chucklu** [comment]
@pavel-mikula-sonarsource  Let's just forget about S3887 , the original code has nothing to do with it.

```C#
public static List<string> MandatoryFields = new List<string>()
        {
            "First name",
            "Last name"
        };
```
The original code currently was reported as https://rules.sonarsource.com/csharp/RSPEC-2223 Non-constant static fields should not be visible.  And the solution of RSPEC-2223 is make the field readonly or const, it's the wrong solution, because it did not recognized the issue correctly.

And it should be recognized as https://rules.sonarsource.com/csharp/RSPEC-2386, Mutable fields should not be "public static".
And the following solution of 2386 works,

> Exceptions
> 
> No issue is reported:
> 
>     If the type of the field inherits/implements one (at least) of the following types:
>         System.Collections.ObjectModel.ReadOnlyCollection<T>
>         System.Collections.ObjectModel.ReadOnlyDictionary<TKey, TValue>
>         System.Collections.Immutable.IImmutableArray<T>
>         System.Collections.Immutable.IImmutableDictionary<TKey, TValue>
>         System.Collections.Immutable.IImmutableList<T>
>         System.Collections.Immutable.IImmutableSet<T>
>         System.Collections.Immutable.IImmutableStack<T>
>         System.Collections.Immutable.IImmutableQueue<T>
>     If the field is readonly and is initialized inline with an immutable type (i.e. inherits/implements one of the types in the previous list) or null.

**Comment by pavel-mikula-sonarsource** [comment]
This example
```
public static List<string> MandatoryFields = new List<string>()
{
    "First name",
    "Last name"
};
```
raises both, S2223 and S2386. So it behaves as specified.

**Comment by chucklu** [comment]
Currently I tried to use IImmutableList<string> as the field type, it still reported as S2223
```C#
 public static IImmutableList<string> MandatoryFields = ImmutableList.Create(
            "First name",
            "Last name"
        );
```
Maybe I should use ImmutableList<string> as the field type, I will try it on 2020-10-09. It's the national day holiday now in China.

**Comment by chucklu** [comment]
> The S3887 doesn't ask you to remove the readonly keyword. It asks you to Use an immutable collection or reduce the accessibility of the field(s).. You should replace your List<string> with string[] or something like ImmutableList<string>.

_`It asks you to Use an immutable collection`_ Actually I did not realized this solution from the Compliant Solution,
From my experience
```C#
public readonly string[] strings;  // Noncompliant
```
was modified to
```C#
 public string[] strings;
```
What I got is, to fix 3387 we can remove the readonly keword. It's misleading.

It would make the documentation better, if you can add your explanation in document.

>  It asks you to Use an immutable collection or reduce the accessibility of the field(s).. You should replace your List<string> with string[] or something like ImmutableList<string>.

And add the sample
```C#
public readonly List<string> strings;  // Noncompliant
```
was modified to
```C#
 public string[] strings;
```

**Comment by pavel-mikula-sonarsource** [comment]
> Currently I tried to use IImmutableList as the field type, it still reported as S2223
> 
> ```cs
>  public static IImmutableList<string> MandatoryFields = ImmutableList.Create(
>             "First name",
>             "Last name"
>         );
> ```
> 
> Maybe I should use ImmutableList as the field type, I will try it on 2020-10-09. It's the national day holiday now in China.

Do you still have a question about this or is it clear now? Using IImmutableList will solve S2386, but the static instance could be replaced by somebody, so it should still raise S2223.

**Comment by pavel-mikula-sonarsource** [comment]
The S3887 message is quite explicit about the immutable collection:
>Use an immutable collection or reduce the accessibility of this field.

But you're right that the compliant example for it is missing. I've updated the RSPEC. The change will be in the next release.

**Comment by chucklu** [comment]
I will try to fix the S2223 issue when I back to work on 2020-10-09. And I will give feedback here.

**Comment by chucklu** [comment]
@pavel-mikula-sonarsource  I am doing a test for sonarqube test , and It's weird.
 the following code was only recognized as 
[RSPEC-1118](https://rules.sonarsource.com/csharp/tag/design/RSPEC-1118) "Utility classes should not have public constructors" 
and I expect it will report as RSPEC-2223 "Non-constant static fields should not be visible
```C#
  public class CustomHelper
    {
        public static List<string> MandatoryFields = new List<string>
        {
            "First name",
            "Last name"
        };
    }
```

**Comment by chucklu** [comment]
@pavel-mikula-sonarsource  And after I add the static keyword, the following code has no issue. Should this be a solution for https://rules.sonarsource.com/csharp/RSPEC-2223?
```C#
 public static class CustomHelper
    {
        public static List<string> MandatoryFields = new List<string>
        {
            "First name",
            "Last name"
        };
    }
```

**Comment by pavel-mikula-sonarsource** [comment]
When I copy paste your example like this:
```C#
using System.Collections.Generic;

public class CustomHelper
{
    public static List<string> MandatoryFields = new List<string>
        {
            "First name",
            "Last name"
        };
}
```
I get these issues in the file:
* S2386	Use an immutable collection or reduce the accessibility of the field(s) 'MandatoryFields'.	
* S1118	Add a 'protected' constructor or the 'static' keyword to the class declaration.	
* S2223	Change the visibility of 'MandatoryFields' or make it 'const' or 'readonly'.	
* S1104	Make this field 'private' and encapsulate it in a 'public' property.
* S3903	Move 'CustomHelper' into a named namespace.

Rule S2223 is triggered as expected. Adding a `static` keyword to the class resolves only S1118, it has no effect on S2223.
